src/scheduler/job_info.c

```diff
@@ -1896,6 +1893,467 @@ translate_fail_code(schd_error *err, char *comment_msg, char *log_msg)
+		return -1;
+
+	for (i = 0; rsets[i] != NULL; i++) {
+		if (qinfo != NULL && rsets[i]->qinfo == NULL || qinfo == NULL && rsets[i]->qinfo != NULL)
+			continue;
+		if (qinfo != NULL && rsets[i]->qinfo != NULL && cstrcmp(qinfo->name, rsets[i]->qinfo->name) != 0)
+			continue;
+		if (user != NULL && cstrcmp(user, rsets[i]->user) != 0)
+			continue;
+		if (group != NULL && cstrcmp(group, rsets[i]->group) != 0)
+			continue;
+		if (project != NULL && cstrcmp(project, rsets[i]->project) != 0)
+			continue;
+		if (sel != NULL && compare_selspec(rsets[i]->select_spec, sel) == 0)
+			continue;
+		if (pl != NULL && compare_place(rsets[i]->place_spec, pl) == 0)
+			continue;
+		if (req != NULL && compare_resource_req_list(rsets[i]->req, req, policy->equiv_class_resdef) == 0)
+			continue;
+
+		/* If we got here, we have found our set */
+		return i;
+	}
+	return -1;
+	
+}
+
+/**
+ * @brief find the index of a resresv_set by a resresv inside it
+ * @param[in] policy - policy info
+ * @param[in] rsets - resresv_set array to search
+ * @param[in] resresv - resresv to search for
+ * @return index of resresv
+ */
+int
+find_resresv_set_by_resresv(status *policy, resresv_set **rsets, resource_resv *resresv)
+{
+	char *user = NULL;
+	char *grp = NULL;
+	char *proj = NULL;
+	queue_info *qinfo = NULL;
+
+	if (policy == NULL || rsets == NULL || resresv == NULL)
+		return -1;
+
+	if (resresv_set_use_user(resresv->server))
+		user = resresv->user;
+
+	if (resresv_set_use_grp(resresv->server))
+		grp = resresv->group;
+
+	if (resresv_set_use_proj(resresv->server))
+		proj = resresv->project;
+
+	if (resresv->is_job && resresv->job != NULL)
+		if (resresv_set_use_queue(resresv->job->queue))
+			qinfo = resresv->job->queue;
+
+	return find_resresv_set(policy, rsets, user, grp, proj, resresv->select, resresv->place_spec, resresv->resreq, qinfo);
+}
+
+/**
+ * @brief create equivalence classes based on an array of resresvs
+ * @param[in] policy - policy info
+ * @param[in] sinfo - server universe
+ * @return array of equivalence classes (resresv_sets)
+ */
+resresv_set **
+create_resresv_sets(status *policy, server_info *sinfo)
+{
+	int i;
+	int j = 0;
+	int cur_ind;
+	int len;
+	resource_resv **resresvs;
+	resresv_set **rsets;
+	resresv_set **tmp_rset_arr;
+	resresv_set *cur_rset;
+
+	if (policy == NULL || sinfo == NULL)
+		return NULL;
+
+	resresvs = sinfo->jobs;
+
+	len = count_array((void **) resresvs);
+	rsets = malloc((len + 1) * sizeof(resresv_set));
+	if (rsets == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		return NULL;
+	}
+
+	rsets[0] = NULL;
+
+	for (i = 0; resresvs[i] != NULL; i++) {
+		cur_ind = find_resresv_set_by_resresv(policy, rsets, resresvs[i]);
+
+		/* Didn't find the set, create it.*/
+		if (cur_ind == -1) {
+			cur_rset = create_resresv_set_by_resresv(policy, sinfo, resresvs[i]);
+			if (cur_rset == NULL) {
+				free_resresv_set_array(rsets);
+				return NULL;
+			}
+			cur_rset->resresv_arr = malloc((len + 1) * sizeof(resource_resv));
```

Comment:
I look at this as the both are about equivalent.  If we go with the way you suggest, we'll constantly be reallocing space to get more when we run out.  If we go with how I've implemented it, we tidy up at the end.  It all depends where we fall between best case and worst case.  If we do what you want the best case will still cause us to realloc multiple times.  The best case is where all the jobs fall in one class.  We'd alloc once for the resresv_sets and then many times as we kept adding jobs to the set.  The worst case is num_jobs resresv_sets with one job per set.  Here we're constantly reallocing resresv_sets, but not for jobs.  If you really want me to change this, I will, but I don't see it as necessary.